### PR TITLE
AST Checklist - Part 2

### LINKED ISSUE #527: AST Fixes checklist
After going through every line in [this](https://github.com/TilakMaddy/foundry-compilers/blob/aderyn/fix/ast-updates/src/artifacts/ast/mod.rs) file and contrasting them against the AST in Aderyn, I have created a dump of `not ok` statements. We need to try to fix each instance on an individual level

```
90:    Block, // notok
91:    Break, // notok
92:    Continue, // notok
93:    DoWhileStatement, // notok
99:    PlaceholderStatement, // notok
124:    TypeName; // notok (i think because of String(String))
174:        canonical_name: Option<String>, // notok
186:        used_events: Vec<usize>, // notok
188:        internal_function_ids: BTreeMap<String, usize>, // notok
209:        base_name: UserDefinedTypeNameOrIdentifierPath, // notok
353:        type_name: ElementaryOrRawTypeName, // notok
360:pub enum ElementaryOrRawTypeName { // notok
428:        index_expression: Option<Expression>, // notok
465:    UnicodeString, // notok
549:impl VariableDeclaration { // notok
576:        overrides: Vec<UserDefinedTypeNameOrIdentifierPath>, // notok
584:        contract_scope: Option<String>, // notok // TODO
676:        error_selector: Option<String>, // notok // TODO
688:        event_selector: Option<String>, // notok // TODO
716:        kind: Option<FunctionKind>, // notok
722:        state_mutability: Option<StateMutability>, // notok
724:        is_virtual: bool, // notok
731:        is_constructor: bool, // notok
738:        is_declared_const: bool, // notok
745:        is_payable: bool, // notok
797:        documentation: Option<String>, // notok
815:    struct DoWhileStatement { // notok
841:        loop_expression: Option<ExpressionStatement>, // notok
873:        external_references: Vec<ExternalInlineAssemblyReference>, // notok
875:        flags: Vec<InlineAssemblyFlag>, // notok
882:pub struct ExternalInlineAssemblyReference { // notok
899:pub enum AssemblyReferenceSuffix {                // notok
912:    MemorySafe, // notok
954:    struct UncheckedBlock { // notok
955:        statements: Vec<Statement>, // notok 
973:        modifier_name: IdentifierOrIdentifierPath, // notok
994:        base_modifiers: Vec<usize>, // notok
1079:    pub foreign: Identifier, // notok 

```

Note To self -
Fix each of the above instance

### PR DESCRIPTION
**Pull request by TilakMaddy**
Pre-requisite: Merge #537 

Adhoc fixes only such as (but not limited to)

* Undeserialized field: Event selector names 
* Undeserialized field: Error selector names
* Undeserialized field: Internal function IDs, etc

Other fixes
* Removed Unhandled invariant in the `Expression` enum

Partially resolves #527 

(I will need 2 more parts to complete the AST fixes but due to the newly assigned priorities in the issues, I will postpone this work in favor of `cyfrinup`)

### CODE REVIEW

**Review by alexroan — CHANGES_REQUESTED** [review]
@TilakMaddy I get an error on (a modified version of) [this codebase](https://github.com/sudoswap/lssvm2) that doesn't occur on `dev` branch.

Trace:

```
thread 'main' panicked at /Users/alexroan/Documents/Cyfrin/aderyn/aderyn_core/src/ast/source_units.rs:47:46:
called `Result::unwrap()` on an `Err` value: Error("missing field `referencedDeclaration`", line: 0, column: 0)
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: <aderyn_core::ast::source_units::SourceUnitNode as serde::de::Deserialize>::deserialize
   4: <serde_json::de::SeqAccess<R> as serde::de::SeqAccess>::next_element_seed
   5: <serde::de::impls::<impl serde::de::Deserialize for alloc::vec::Vec<T>>::deserialize::VecVisitor<T> as serde::de::Visitor>::visit_seq
   6: <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_seq
   7: <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_struct
   8: serde_json::de::from_trait
   9: aderyn_driver::process_auto::absorb_ast_content_into_context
  10: aderyn_driver::process_auto::with_project_root_at::{{closure}}
  11: rayon::iter::plumbing::Folder::consume_iter
  12: rayon::iter::plumbing::bridge_producer_consumer::helper
  13: <rayon::vec::IntoIter<T> as rayon::iter::IndexedParallelIterator>::with_producer
  14: rayon::iter::extend::<impl rayon::iter::ParallelExtend<T> for alloc::vec::Vec<T>>::par_extend
  15: aderyn_driver::process_auto::with_project_root_at
  16: aderyn_driver::driver::drive_with
  17: aderyn_driver::driver::drive
  18: aderyn::main
```

Could there still be an UnhandledExpression that we're not accounting for?

**Comment by TilakMaddy** [comment]
> @TilakMaddy I get an error on (a modified version of) [this codebase](https://github.com/sudoswap/lssvm2) that doesn't occur on `dev` branch.
> 
> Trace:
> 
> ```
> thread 'main' panicked at /Users/alexroan/Documents/Cyfrin/aderyn/aderyn_core/src/ast/source_units.rs:47:46:
> called `Result::unwrap()` on an `Err` value: Error("missing field `referencedDeclaration`", line: 0, column: 0)
> stack backtrace:
>    0: _rust_begin_unwind
>    1: core::panicking::panic_fmt
>    2: core::result::unwrap_failed
>    3: <aderyn_core::ast::source_units::SourceUnitNode as serde::de::Deserialize>::deserialize
>    4: <serde_json::de::SeqAccess<R> as serde::de::SeqAccess>::next_element_seed
>    5: <serde::de::impls::<impl serde::de::Deserialize for alloc::vec::Vec<T>>::deserialize::VecVisitor<T> as serde::de::Visitor>::visit_seq
>    6: <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_seq
>    7: <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_struct
>    8: serde_json::de::from_trait
>    9: aderyn_driver::process_auto::absorb_ast_content_into_context
>   10: aderyn_driver::process_auto::with_project_root_at::{{closure}}
>   11: rayon::iter::plumbing::Folder::consume_iter
>   12: rayon::iter::plumbing::bridge_producer_consumer::helper
>   13: <rayon::vec::IntoIter<T> as rayon::iter::IndexedParallelIterator>::with_producer
>   14: rayon::iter::extend::<impl rayon::iter::ParallelExtend<T> for alloc::vec::Vec<T>>::par_extend
>   15: aderyn_driver::process_auto::with_project_root_at
>   16: aderyn_driver::driver::drive_with
>   17: aderyn_driver::driver::drive
>   18: aderyn::main
> ```
> 
> Could there still be an UnhandledExpression that we're not accounting for?

Interesting... Let me look into it! Can you confirm that the current `dev` works okay on that modified version of `lssvm2`?

**Comment by TilakMaddy** [comment]
I am more inclined towards finding out what the node is missing / corrupted and only If I fail to do that, I would want to bring back Unhandled clause

**Comment by TilakMaddy** [comment]
@alexroan Can you retry on https://github.com/Cyfrin/aderyn/pull/544 ? (It's not a fix but it will tell me where exactly in the deserialization path it fails)

**Comment by alexroan** [comment]
New error here @TilakMaddy (full trace):

```
thread 'main' panicked at aderyn_driver/src/process_auto.rs:165:73:
called `Result::unwrap()` on an `Err` value: Error("missing field `referencedDeclaration`", line: 1, column: 2542)
stack backtrace:
   0:        0x10136601c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf8d878aec918681b
   1:        0x101387b08 - core::fmt::write::h5cb3cc404c6130ab
   2:        0x101362224 - std::io::Write::write_fmt::hd3451e73819d4e5d
   3:        0x101365e74 - std::sys_common::backtrace::print::h33a85a6b603d93c4
   4:        0x1013670bc - std::panicking::default_hook::{{closure}}::hd4e10babaa2d77e7
   5:        0x101366db0 - std::panicking::default_hook::ha2368c966c1e134e
   6:        0x101367978 - std::panicking::rust_panic_with_hook::h8802696766858834
   7:        0x1013673a4 - std::panicking::begin_panic_handler::{{closure}}::h7d9285b392744c48
   8:        0x1013664a0 - std::sys_common::backtrace::__rust_end_short_backtrace::h66844ca21338b1bd
   9:        0x101367114 - _rust_begin_unwind
  10:        0x1013d0194 - core::panicking::panic_fmt::hc3e7edb051023bb5
  11:        0x1013d0508 - core::result::unwrap_failed::h9a52ded9b89ecb04
  12:        0x100d6f6f8 - aderyn_driver::process_auto::absorb_ast_content_into_context::h2f22a6d9c7d39bd5
  13:        0x100d6e56c - aderyn_driver::process_auto::with_project_root_at::{{closure}}::hf7df90125be7258d
  14:        0x100d654f4 - rayon::iter::plumbing::Folder::consume_iter::h8286dbb20bf8a79a
  15:        0x100d24a8c - rayon::iter::plumbing::bridge_producer_consumer::helper::h2c42f4c89e736bb7
  16:        0x100d259b0 - rayon::iter::plumbing::bridge::h815f32e2c588dbf3
  17:        0x100d24230 - rayon::iter::extend::<impl rayon::iter::ParallelExtend<T> for alloc::vec::Vec<T>>::par_extend::h2a9105caa531f691
  18:        0x100d6c3e4 - aderyn_driver::process_auto::with_project_root_at::h80f6d98d50ddc309
  19:        0x100d68ca8 - aderyn_driver::driver::drive_with::hd37084ac40d8e3c7
  20:        0x100d6754c - aderyn_driver::driver::drive::haf479d6fc307fe78
  21:        0x100b206d0 - aderyn::main::h752cfd7276071af7
  22:        0x100b19814 - std::sys_common::backtrace::__rust_begin_short_backtrace::h3254df2540cb9ce4
  23:        0x100b19a30 - std::rt::lang_start::{{closure}}::ha54baf77fa98f997
  24:        0x10135b6f4 - std::rt::lang_start_internal::hb7d45ec431a51272
  25:        0x100b25fa8 - _main
```

**Comment by TilakMaddy** [comment]
> New error here @TilakMaddy (full trace):
> 
> ```
> thread 'main' panicked at aderyn_driver/src/process_auto.rs:165:73:
> called `Result::unwrap()` on an `Err` value: Error("missing field `referencedDeclaration`", line: 1, column: 2542)
> stack backtrace:
>    0:        0x10136601c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf8d878aec918681b
>    1:        0x101387b08 - core::fmt::write::h5cb3cc404c6130ab
>    2:        0x101362224 - std::io::Write::write_fmt::hd3451e73819d4e5d
>    3:        0x101365e74 - std::sys_common::backtrace::print::h33a85a6b603d93c4
>    4:        0x1013670bc - std::panicking::default_hook::{{closure}}::hd4e10babaa2d77e7
>    5:        0x101366db0 - std::panicking::default_hook::ha2368c966c1e134e
>    6:        0x101367978 - std::panicking::rust_panic_with_hook::h8802696766858834
>    7:        0x1013673a4 - std::panicking::begin_panic_handler::{{closure}}::h7d9285b392744c48
>    8:        0x1013664a0 - std::sys_common::backtrace::__rust_end_short_backtrace::h66844ca21338b1bd
>    9:        0x101367114 - _rust_begin_unwind
>   10:        0x1013d0194 - core::panicking::panic_fmt::hc3e7edb051023bb5
>   11:        0x1013d0508 - core::result::unwrap_failed::h9a52ded9b89ecb04
>   12:        0x100d6f6f8 - aderyn_driver::process_auto::absorb_ast_content_into_context::h2f22a6d9c7d39bd5
>   13:        0x100d6e56c - aderyn_driver::process_auto::with_project_root_at::{{closure}}::hf7df90125be7258d
>   14:        0x100d654f4 - rayon::iter::plumbing::Folder::consume_iter::h8286dbb20bf8a79a
>   15:        0x100d24a8c - rayon::iter::plumbing::bridge_producer_consumer::helper::h2c42f4c89e736bb7
>   16:        0x100d259b0 - rayon::iter::plumbing::bridge::h815f32e2c588dbf3
>   17:        0x100d24230 - rayon::iter::extend::<impl rayon::iter::ParallelExtend<T> for alloc::vec::Vec<T>>::par_extend::h2a9105caa531f691
>   18:        0x100d6c3e4 - aderyn_driver::process_auto::with_project_root_at::h80f6d98d50ddc309
>   19:        0x100d68ca8 - aderyn_driver::driver::drive_with::hd37084ac40d8e3c7
>   20:        0x100d6754c - aderyn_driver::driver::drive::haf479d6fc307fe78
>   21:        0x100b206d0 - aderyn::main::h752cfd7276071af7
>   22:        0x100b19814 - std::sys_common::backtrace::__rust_begin_short_backtrace::h3254df2540cb9ce4
>   23:        0x100b19a30 - std::rt::lang_start::{{closure}}::ha54baf77fa98f997
>   24:        0x10135b6f4 - std::rt::lang_start_internal::hb7d45ec431a51272
>   25:        0x100b25fa8 - _main
> ```

For #545 ??

**Comment by alexroan** [comment]
Nope, that was in this branch using the latest commit.

File: aderyn_core/src/ast/literals.rs [hunk 1]
```diff
@@ -12,6 +12,7 @@ pub enum LiteralKind {
     String,
     HexString,
     Address,
+    UnicodeString, // notok
```
Comment (alexroan) [inline aderyn_core/src/ast/literals.rs:15]: `// notok` ??

Comment (TilakMaddy) [reply aderyn_core/src/ast/literals.rs:15]: I copied that whole thing by mistake. (It was from my personal notes)

Comment (TilakMaddy) [reply aderyn_core/src/ast/literals.rs:15]: It is okay !

**Pull request merged**